zebra-consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,158 @@
+                return async { Err("the block does not have a coinbase height".into()) }.boxed()
+            }
+        };
+
+        // TODO(teor):
+        //   - implement chaining from checkpoints to their ancestors
+        //   - if chaining is expensive, move this check to the Future
+        //   - should the state contain a mapping from previous_block_hash to block?
+        let checkpoint_hash_ref = match checkpoints.get(&block_height) {
+            Some(hash) => hash,
+            None => {
+                return async { Err("the block's height is not a checkpoint height".into()) }
+                    .boxed()
+            }
+        };
+
+        // Avoid moving a reference into the future.
+        let checkpoint_hash = *checkpoint_hash_ref;
+
+        // `state_service.call` is OK here because we already called
+        // `state_service.poll_ready` in our `poll_ready`.
+        let add_block = self.state_service.call(zebra_state::Request::AddBlock {
+            block: block.clone(),
+        });
+
+        async move {
+            // Hashing is expensive, so we do it in the Future
+            if BlockHeaderHash::from(block.as_ref()) != checkpoint_hash {
+                // The block is on a side-chain
+                return Err("the block hash does not match the checkpoint hash".into());
+            }
+
+            match add_block.await? {
+                zebra_state::Response::Added { hash } => Ok(hash),
+                _ => Err("adding block to zebra-state failed".into()),
+            }
+        }
+        .boxed()
+    }
+}
+
+// TODO(teor): add a function for the maximum checkpoint height
+// We can pre-calculate the result in init(), if we want.
+
+/// Return a checkpoint verification service, using the provided state service.
+///
+/// The checkpoint verifier holds a state service of type `S`, into which newly
+/// verified blocks will be committed. This state is pluggable to allow for
+/// testing or instrumentation.
+///
+/// The returned type is opaque to allow instrumentation or other wrappers, but
+/// can be boxed for storage. It is also `Clone` to allow sharing of a
+/// verification service.
+///
+/// This function should be called only once for a particular state service (and
+/// the result be shared) rather than constructing multiple verification services
+/// backed by the same state layer.
+pub fn init<S>(
+    state_service: S,
+    checkpoint_list: Option<HashMap<BlockHeight, BlockHeaderHash>>,
+) -> impl Service<
+    Arc<Block>,
+    Response = BlockHeaderHash,
+    Error = Error,
+    Future = impl Future<Output = Result<BlockHeaderHash, Error>>,
+> + Send
+       + Clone
+       + 'static
+where
+    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
+        + Send
+        + 'static,
+    S::Future: Send + 'static,
+{
+    Buffer::new(
+        CheckpointVerifier {
+            state_service,
+            checkpoint_list,
+        },
+        1,
+    )
+}
```

Comment:
tioli: if we're not planning to expose the checkpoint verifier as part of the public API, rather than hiding it behind an `init` function that wraps it in a buffer, we could have the rest of the crate use it directly.

Reply:
Thanks for that reminder, I'll stop it being public.

Do we always want each verifier wrapped in a buffer?

Reply:
I don't think it's necessary to wrap a verifier in a buffer unless we know that we want to share the verifier, since it adds some overhead and extra complexity.  We might only need to share a highest-level verifier that we return from the crate's public `init()` method, and have that verifier own component verifiers like this one directly.

Reply:
Do we want zebra-consensus users to be able to use BlockVerifier directly? Or just HybridVerifier?

We can let HybridVerifier use BlockVerifier directly, but require outside users to init() them both with buffers.

Reply:
Not sure! We can always leave things public for now and then clean up the API later, moving stuff around is relatively easy.  When we wrote `zebra_network` we didn't get the nice encapsulated API until the very end with a pass to clean up imports.

Reply:
Fixed in ea54df91fce8e644e1060d1541ba4e0e6be0b128, but I left init() public for now.